Revision of the Context documentation (#6008)
typst/typst: docs/reference/language/context.md

[Show rules]($styling/#show-rules) provide context[^1] and numberings in the
outline, for instance, also provide the proper context to resolve counters.

## Behavior of the context keyword
Style properties frequently change within a document, for example by applying set 
rules. To retrieve such poperties in a consistent way, one must first specify 

> the context where the query is to be executed. This is the purpose of the 

Reviewer: the term "query" has a rather specific connotation in Typst and should be avoided here imo (also a bunch of times more below)

Author: I deliberately use `query` when I refer to the specific typst meaning, and plain text to just use the English term. IMHO, there is no confusion. Anyway, what synonymous word would not collide with anything in your opinion?

Reviewer: I think "retrieve" is good. The difference between "query" and `query` is a bit too subtle imo.

> `context` keyword. Once the context has been fixed, the property information 

Reviewer: Saying "This is the purpose of the context keyword" is a bit strong since it's not the only purpose of it

Author: Please enlighten me about other purposes (or give me a pointer).

Reviewer: Context also provides the current location via `here()`.

> is available through a simple field access syntax. For example, `text.lang` 

Reviewer: ```suggestion
is available through field access syntax. For example, `text.lang` 
```

Author: You should be proud of how simple it is (in comparison to the competition :-).

Reviewer: :D I'm trying to reduce the use of "simple" in my writing as things are differently simple/hard for different people and it's typically unnecessary fluff.

asks for the current language setting. In its simplest form, the `context` 
keyword refers to "right here":

```example
#set text(lang: "de")
// query the language setting "here"
#context text.lang
```

> Note that calling `#text.lang` directly would be an error, because the request

Reviewer: "calling" isn't really the correct term I think, as it's not a function (this also appears various times more below)

Author: I'm running out of words -- "query" is also forbidden :-) How about "retrieving"?

Reviewer: Retrieving is good.

cannot be answered without knowledge of the context. The field names supported 
by a given element function always correspond to the named parameters documented 
on each element's page.

Moreover, some functions, such as [`to-absolute`]($length.to-absolute) 
and [`counter.display`]($counter.display) are only applicable in a context, 
because their results depend on the current settings of style properties. 
When another function `foo()` calls a context-dependent function, it becomes 
itself context-dependent:

```example
#let foo() = 1em.to-absolute()
#context {

>   // foo() cannot be called outside of a context

Reviewer: if managable, please try to keep examples to <=38 columns so that they will display side-by-side in the docs and are easy to read on mobile

Author: Thanks. Was wondering what the maximum width is...

  foo() == text.size
}
```

When a property is changed, the response to the query changes accordingly: 

```example
#set text(lang: "en")
#context text.lang

#set text(lang: "de")
#context text.lang
```

> The output of a `#context ...` call is static in the form of opaque 

Reviewer: ```suggestion
The value of a `#context ...` expression is static in the form of opaque 
```

Author: How about "The output of a context expression"?

See also my remark below questioning if the value/output of a context expression is really always `content`.

[`content`]. Write access to context output is prohibited, as it would 
often result in invalid code: If the context changes between read and 

> write, overwriting a property would cause an inconsistent system state. 

Reviewer: This doesn't read quite correctly. The crucial thing is that contextual content might appears 0, 1, or N times in the document. There is some more discussion of this in the second part of this blog post: https://laurmaedje.github.io/posts/types-and-context/

Author: I'm not happy with this sentence either. However, I would like to have a one-sentence explanation why things need to be more complicated than what one would naively expect. Talking about "closures" etc. as in your post doesn't feel right, because many users won't know this term. Any suggestions?

In fact, context-dependent property fields are immutable constants even 
within the context itself:

```example
#set text(lang: "en")
#context [

>   call 1: #text.lang \

Reviewer: ```suggestion
  call 1: #text.lang
```
there is already a paragraph break

  #set text(lang: "fr")
  call 2: #text.lang
]
```

> Both calls have the same output 'en', because `text.lang` is assigned

Reviewer: ```suggestion
Both calls have the same output "en", because `text.lang` is assigned
```
We don't really use single quotes except in nested quote situations.

Author: I gave up enforcing so detailed style guides in my OS projects -- was just not worth the trouble.

upon entry in the context and remains constant until the end of its scope 
(the closing `]`). It does not "see" the `#set text(lang: "fr")` before 
call 2. Compare this to the previous example: there we got two different 
results because we created two different contexts.

However, immutability only applies to the property fields themselves. 

> Content creation instructions within a context _do_ see the effect of 

Reviewer: "Content creation instructions" is a bit of a mouthful, maybe just "content expressions"

Author: I feel that "Content creation instructions" is a less technical term, so maybe better for newcomers.

Reviewer: "Content expressions" is indeed bad. I think my mind was thinking about "context expressions" when I wrote this. But the new version in the latest commit reads well.

the set rule. Consider the same example with font size:

```example
#set text(size: 50pt)
#context [
  call 1: #text.size \

  #set text(size: 25pt)
  call 2: #text.size
]
```

Call 2 still outputs '50pt', because `text.size` is a constant.
However, this output is printed in '25pt' font, as specified by the set
rule before the call. This illustrates the importance of picking the 
right insertion point for a context to get access to precisely the right 
styles. If you need access to updated property fields after a set rule, 
you can use nested contexts:

```example
#set text(lang: "en")
#context [
  call 1: #text.lang \

  #set text(lang: "fr")
  call 2: #context text.lang
]
```

All of the above applies to `show` rules analogously, for example:

```example
#let template(body) = {
  set text(size: 25pt)
  body
}

#set text(size: 50pt)
#context [
  call 1: #text.size \

  #show: template
  call 2: #text.size \
  call 3: #context text.size
]
```

> ## Controlling content creation within a context

Reviewer: I'm not quite sure what this title means

Author: It meant to say that a `context` provides a lot more power to configure and control content creation programmatically because one has access to the parameter fields. The section is supposed to demonstrate this.

Author: I should definitely mention this example:
```typ
// make text darker
#set text(fill: context text.fill.darken(20%))
```

Author: However, on first glance this example contradicts the statement that the value of a `context` is `content`: Here, it seems to be a ytpe that can be assigned to the fill parameter. For example, this does not work
```typ
#set text(fill: [something])
```
So, clearly the output of `context text.fill.darken(20%)` is not `content`.

Reviewer 2: The value of `text.fill.darken(20%)` is not content, but the value of `context text.fill.darken(20%)` is.

Reviewer: > I should definitely mention this example:

This example doesn't work yet, it was just a proposed idea, so it shouldn't be mentioned in the docs.

The main purpose of retrieving the current values of properties is, 
of course, to use them in the calculation of derived properties, 

> instead of setting those properties manually. For example, you can

Reviewer: I wouldn't say that's accurate. It's surely one use case, but you might just as well use it for something else.

Reviewer: For example to localize a document via `context if text.lang == "de" [Inhalt] else [Contents]`. I think the use cases are very diverse, which is why I stumbled over the "of course"

double the font size like this:

```example
#context {
  // the context allows you to
  // retrieve the current text.size
  set text(size: text.size * 200%)
  [large text \ ] 
}
original size
```

Since set rules are only active until the end of the enclosing scope, 
'original size' is printed with the original font size.
The above example is equivalent to 

```example
#{
  set text(size: 2em)
  [large text \ ]
}

> original size

Reviewer: The start of a phrase in example should be capitalized. E.g. "Original size" or "Call 1" (bad example because call is not the correct term, but just an example).

```

but convenient alternatives like this do not exist for most properties.
This makes contexts a powerful and versatile concept. For example, 
to double the spacing between the lines of an equation block, you can 
use the same resizing technique in a show rule. In this case, explicitly
adding the `context` keyword is not necessary, because a show rule
establishes a context automatically:

```example
#let spaced-eq(spacing: 100%, body) = {
  show math.equation.where(block: true): it => {
    // access current par.leading in the
    // context of the show rule
    set par(leading: par.leading * spacing)
    it
  }
  body
}

normal spacing:
$
x \
x
$
doubled spacing:
#spaced-eq(spacing: 200%)[$
z \
z

> $]

Reviewer: This example is a bit to complex for my liking. I think the point can be made in a clearer way.

Author: Would it be better to write `$ x \ x $` ?

[Show rules]($styling/#show-rules) provide context[^1] and numberings in the
outline, for instance, also provide the proper context to resolve counters.

## Behavior of the context keyword

> Style properties frequently change within a document, for example by applying set 

Reviewer: "by applying set rules" reads a bit clunkily to me. maybe "Style properties frequently change within a document, through use of set rules"

> rules. To retrieve such properties in a consistent way, one must first specify 

Reviewer: not sure whether "consistent" is quite on point. Maybe "well-defined"?

the precise context where the property should be retrieved. This can be achieved 
with the `context` keyword. Once the context has been fixed, the property 
information is available through standard field access syntax. For example, 
`text.lang` asks for the current language setting. In its simplest form, the 
`context` keyword refers to "right here":

```example
#set text(lang: "de")
// read the language setting "here"
#context text.lang

> ```

Reviewer 3: In fact, we could consider mentioning this issue very early on below, when saying that context returns content, if possible. (Needs further discussion.)

> ```

Reviewer 3: Actually, this is already _briefly_ mentioned at the top, but either way people generally skim the docs for examples they may paste elsewhere (for good reason), so I'd argue this is still a concern, and an explicit (counter-)example would be nice.

> ```

Author: This example is the "Hello world" of contexts. Everyone knows that "Hello world" is not very useful, but it gets across a few very basic things (such as how the context keywords looks like, that properties are accessed by field syntax, and that they can be displayed). Of course, a different example (instead of `lang`) might be more instructive.

> ```

Author: BTW, the example is not completely pointless -- it is useful as a quick-and-dirty way of inspecting what the current value of some property is.

> ```

Author: A remark about the error message when a field is accessed outside of a context:
```typ
#text.lang    // => "Can only be used when context is known."
```
This message is a mystery for beginners -- when you haven't heard about the `context` concept and read "context is known" with its generic, everyday-language meaning, you have no idea what to do (and this unfortunately applies to many other error messages as well). Worse, when I first opened the context documentation, I got the feeling that this is meant for advanced users and wouldn't be needed right now. In reality, contexts are absolutely fundamental, and an improved documentation should get this across.

> ```

Reviewer 3: > This example is the "Hello world" of contexts. Everyone knows that "Hello world" is not very useful, but it gets across a few very basic things (such as how the context keywords looks like, that properties are accessed by field syntax, and that they can be displayed)

Yeah of course, but as someone who is often active in the typst support channels, i find the problem I mentioned to be rather common across beginners. See also https://forum.typst.app/t/why-is-the-value-i-receive-from-context-always-content/164/

It is possible to keep the example in some way, but maybe it just shouldn't be the _first_.

> ```

Reviewer: It's true that it's a very common problem with beginners, but I'm not sure that this specific example is the problem. No matter what you put there, people will think that's the way it works. So I think we can keep that. But I agree that we should have more explanation specific to that problem somewhere relatively early in the docs. Doesn't necessarily have to be part of this PR though.

Note that any attempt to access `#text.lang` directly, i.e. outside of a context,

> will cause the compiler to issue an error message. The field names supported 

Reviewer: I think it's worth adding an extra sentence here explaining why the compiler has no choice but the emit an error: Because it needs to know at which location you want the property.

by a given element function always correspond to the named parameters documented 
on each element's page.

Moreover, some functions, such as [`to-absolute`]($length.to-absolute) 
and [`counter.display`]($counter.display), are only applicable in a context, 
because their results depend on the current settings of style properties. 
When another function `foo()` calls a context-dependent function, it becomes 
itself context-dependent:

```example
#let foo() = 1em.to-absolute()
#context {
  // foo() cannot be called
  // outside of a context
  foo() == text.size
}
```

When a property is changed, the response to the property access 
changes accordingly: 

```example
#set text(lang: "en")
#context text.lang

#set text(lang: "de")
#context text.lang
```

As you see, the result of a `#context ...` expression can 
be inserted into the document as `content`. Context blocks can 
contain arbitrary code beyond the field access. However,

> and this is often surprisingly for newcomers, context-dependent 

Reviewer: ```suggestion
and this is often surprising for newcomers, context-dependent 
```

property fields remain _constant_ throughout the context's scope. 
This has two important consequences: First, direct property 
assignments like `text.lang = "de"` are _not_ allowed &ndash; 
use `set` or `show` rules for this purpose. Second, changes to a 
property value within a context (e.g. by a `set` rule) are not 

> observable by field access within that same context:

Author: > This mention to property assignment (at least here) is confusing, since that was never suggested to be possible in the first place

The problem with your argument is that anyone with the least programming experience assumes by default that properties can be assigned -- it is just how it works in most languages. That typst never suggested the possibility is irrelevant, the brain automatically relates new stuff to what it already knows.

In fact, a major reason why I struggled with the typst documentation (and take the time to propose improvements) is that one has to unlearn various things that other programming languages take for granted. You absolutely must mention and explain when typst deviates from the mainstream.

> observable by field access within that same context:

Reviewer 3: Yes, I agree this should be mentioned, but my main point is that this isn't related to `context` itself and feels out of place. This is rather related to `set` rules themselves - that's how you set properties on elements, not through assignment.

In other words, since `set` rules from outside are what define the value of `text.lang` in the first place, one could naturally expect that a `set` rule inside `context { ... }` would also change it within. That is not the case.

As a compromise, the suggestion of not being able to assign to `text.lang` could be moved to a note after the phrase or paragraph, e.g. "set rules cannot change `text.lang` within the context (nor can you assign to them, as only set rules can influence the default values of element properties)."

> observable by field access within that same context:

Author: > In other words, since set rules from outside are what define the value of text.lang in the first place, one could naturally expect that a set rule inside context { ... } would also change it within. That is not the case.

Isn't this exactly what my proposed documentation says?

> observable by field access within that same context:

Reviewer 3: > Isn't this exactly what my proposed documentation says?

I'm effectively proposing for that to be the _only_ thing it says, and mention the assignment thing elsewhere, or as a sidenote.

```example
#set text(lang: "en")
#context [
  Read 1: #text.lang

  #set text(lang: "fr")
  Read 2: #text.lang
]
```

Both reads have the same output `"en"`, because `text.lang` is assigned
upon entry in the context and remains constant until the end of its scope 

> (the closing `]`). Thus, the `text.lang` field is not affected by

Reviewer: ```suggestion
(the closing `]`). Thus, the second `text.lang` access is not affected by
```

> #set text(lang: "fr")`, although Read 2 occurs afterwards. Compare 

Reviewer: ```suggestion
`[#set text(lang: "fr")]`, although Read 2 occurs after it. Compare 
```

this to the previous example: There we got two different results because 
we created two different contexts.

However, immutability only applies to the property fields themselves. 
The appearance of content within a context _can_ be changed in the 
usual manner, e.g. by set rules. Consider the same example with font size:

```example
#set text(size: 40pt)
#context [
  Read 1: #text.size

  #set text(size: 25pt)
  Read 2: #text.size
]
```

> Read 2 still outputs `40pt`, because `text.size` is a constant.

Reviewer: ```suggestion
Read 2 still outputs `{40pt}`, because `text.size` is a constant.
```

However, this output is printed in 25pt font, as specified by the set
rule before the read. This illustrates the importance of picking the 
right insertion point for a context to get access to precisely the right 
styles. If you need access to updated property fields after a set rule, 
you can use _nested contexts_:

```example
#set text(lang: "en")
#context [
  Read 1: #text.lang

  #set text(lang: "fr")
  Read 2: #context text.lang
]
```

> All of the above applies to `show` rules analogously. To demonstrate this, 

Reviewer 3: Otherwise they don't affect context at all. That's the only way they could potentially change a property (and, indeed, they affect nested contexts). But a show rule transforming element A into B wouldn't have any impact whatsoever.

> All of the above applies to `show` rules analogously. To demonstrate this, 

Reviewer 3: This is mostly due to how the `show` syntax is a bit overloaded. Only "everything show rules" - equivalent to passing the rest of a scope to a function - can also apply show rules on the current scope.

```

Since set rules are only active until the end of the enclosing scope, 
"Original size" is printed with the original font size.
For the specific case of accessing `text.size`, context is usually not necessary as the `1em` unit is always equal to the current font size, so the above example is equivalent to 

```example
#[
  #set text(size: 2em)
  Large text \ 
]
Original size
```

but convenient alternatives like this are unavailable for most properties.
This makes contexts a powerful and versatile concept. For example, 
you can use a similar resizing technique to increase the spacing 
between the lines of a specific equation block (or any other content):

```example
#let spaced(spacing: 100%, body) = context {
  // access current par.leading in a context
  set par(leading: par.leading * spacing)
  body
}

Normal spacing:
$ x \ x $
Doubled spacing:
#spaced(spacing: 200%)[$ z \ z $]
```

The advantage of this technique is that the user does not have to know the 
original spacing in order to double it. To double the spacing of all 
equations, you can put the same calculations in a show rule. Note that 
it is not necessary to add the `context` keyword on the right-hand side 

> of a `show` rule, because show rules establish a context automatically:

Reviewer: The fact that show rules establish context automatically is maybe a bit too buried. It's quite important I think.

```example
Normal spacing:
$ x \ x $

> #show math.equation.where(block: true): it => {

Reviewer: The rendered version is fairly long and noisy now because both examples don't fit side-by-side and they show the exact same output. Maybe we can somehow shorten this? Maybe the second version could be dropped in favor of a new section that explains that show rules establish context and then have a different example there?

<img width="508" alt="image" src="https://github.com/user-attachments/assets/de360c04-e883-474f-a01a-2d868a519698" />

  // access current par.leading in a context,
  // established automatically by the show rule
  set par(leading: par.leading * 200%)
  it
}

Doubled spacing:
$ z \ z $

> ```

Reviewer 3: This has the unfortunate pitfall of not being overridable by a later set rule, since an earlier show rule executes last... it's also why `show: set` rules are usually recommended instead.

Of course, that's not a problem if you never try to override it, but wanted to highlight this for discussion.

Author: You mean the example should be written like this?
```typ
#show math.equation.where(block: true): set par(leading: par.leading * 200%)
```

This is a typical example of a problem I noticed with the current documentation: After reading it, one tends to overcomplicate solutions.

Reviewer: > You mean the example should be written like this?
> ```typ
> #show math.equation.where(block: true): set par(leading: par.leading * 200%)
> ```

Unfortunately, this doesn't work because`par.leading * 200%` is evaluated when the show rule is first processed (without context), not when the equation is visited.

outline, for instance, also provide the proper context to resolve counters.

## Style context
Style properties often change within a document, for example by applying set 
rules. Consequently, to retrieve settings one must first specify the context 
where the query is to be executed. Within a given context, the property 
information is provided by a simple field access syntax. For example, 
`text.lang` asks for the current language setting. In its simplest form, the
`context` keyword refers to "right here":

```example
#set text(lang: "de")
#context text.lang
```

Note that calling `#text.lang` directly would be an error, because the request
cannot be answered without knowledge of the context. The fields supported by

> a given element function are documented **where??** and can be retrieved in a 

Reviewer 2: The fields are just the named parameters documented on each element's page.

Author: Took this remark into my revised version.

> document by calling **what?? (something like `fields()`)**.

Reviewer 2: I don't think there is a way to dynamically get the list of available fields of an element in the general case.

Author: That's sad. Looks natural when you come from Python.

When the language setting changes, the responses to the query change accordingly: 
_Remark: The old example with `#let value = context text.lang` is very confusing at 
this early stage of the explanation and does more harm than good._

```example
#context text.lang

#set text(lang: "de")
#context text.lang

#set text(lang: "fr")
#context text.lang
```

> The output of a `#context ...` call is _read-only_ (in the form of `[content]`).

Reviewer 2: ```suggestion
The output of a `#context ...` call is _read-only_ (in the form of `content`).
```

I'm not sure what you intended here, but I believe the brackets are unnecessary.

Author: They were in the original version, though ...

Reviewer 2: Oh i see. That was to turn it into a link. You can make it code and a link with `` [`content`] `` instead.

Allowing write access would likely result in invalid code, because the context
might have already changed in the meantime. Therefore, temporary changes of
settings must be done within the context, and they are only active until the 
end of the context's scope:

```example
#context {
  // the context allows you to retrieve the current text.size
  set text(size: text.size * 200%)
  [large text] 
}
original size

> ```

Reviewer 2: I think this part could be improved. I'm not sure what you mean by "read-only". What would the user expect to be modifiable that is not?

I like the example, though.

Author: I hope the new version is an improvement.

#context text.lang
```

As you see, the result of a `#context ...` expression can 
be inserted into the document as `content`. Context blocks can 
contain arbitrary code beyond the field access. However,
and this is often surprisingly for newcomers, context-dependent 
property fields remain _constant_ throughout the context's scope. 
This has two important consequences: First, direct property 
assignments like `text.lang = "de"` are _not_ allowed &ndash; 
always use `set` or `show` rules. Second, changes to a 
property value within a context (e.g. by a `set` rule) are not 
observable by field access within that same context:

```example
#set text(lang: "en")
#context [
  Read 1: #text.lang

  #set text(lang: "fr")
  Read 2: #text.lang
]
```

Both reads have the same output `"en"`, because `text.lang` is assigned

> upon entry in the context and remains constant until the end of its scope 

Author: While your explanation may be technically more precise, I think it is harder to understand for newcomers. I don't see why "assigned upon entry into the context" is wrong and would rather like to keep it. May "entry into the context block", but I'm not sure of this is better.

(the closing `]`). Thus, the `text.lang` field cannot "see" the effect 
of `#set text(lang: "fr")`, although Read 2 occurs afterwards. Compare 
this to the previous example: There we got two different results because 
we created two different contexts.

However, immutability only applies to the property fields themselves. 
The appearance of content within a context _can_ be changed in the 
usual manner. e.g. by set rules. Consider the same example with font size:

```example
#set text(size: 40pt)
#context [
  Read 1: #text.size \

  #set text(size: 25pt)
  Read 2: #text.size
]
```

Read 2 still outputs `40pt`, because `text.size` is a constant.
However, this output is printed in 25pt font, as specified by the set
rule before the read. This illustrates the importance of picking the 
right insertion point for a context to get access to precisely the right 
styles. If you need access to updated property fields after a set rule, 
you can use _nested contexts_:

```example
#set text(lang: "en")
#context [
  Read 1: #text.lang \

  #set text(lang: "fr")
  Read 2: #context text.lang
]
```

> All of the above applies to `show` rules analogously. To demonstrate this, 

Author: I have no idea either.

> All of the above applies to `show` rules analogously. To demonstrate this, 

Author: I changed it back to "applies" by majority vote, but Gemini thinks that "apply" was correct after all.

> All of the above applies to `show` rules analogously. To demonstrate this, 

Author: DeepL favors "All of the above applies", Google translate opts for "Everything above applies"

properties, instead of setting those properties manually. For example, 
you can double the font size like this:

```example
#context [
  // the context allows you to
  // retrieve the current text.size
  #set text(size: text.size * 200%)
  Large text \ 
]
Original size
```

Since set rules are only active until the end of the enclosing scope, 
"Original size" is printed with the original font size.
The above example is equivalent to 

```example
#[
  #set text(size: 2em)
  Large text \ 
]
Original size
```

but convenient alternatives like this are unavailable for most properties.
This makes contexts a powerful and versatile concept. For example, 
you can use a similar resizing technique to increase the spacing 
between the lines of a specific equation block (or any other content):

```example
#let spaced(spacing: 100%, body) = context {
  // access current par.leading in a context
  set par(leading: par.leading * spacing)
  body
}

Normal spacing:
$ x \ x $
Doubled spacing:
#spaced(spacing: 200%)[$ z \ z $]
```

The advantage of this technique is that the user does not have to know the 
original spacing in order to double it. To double the spacing of all 

> equations, you can put the same calculations in a `show` rule. Note that 

Reviewer 2: ```suggestion
equations, you can put the same calculations in a show rule. Note that 
```

The documentation usually uses this convention.

The rule that context-dependent variables and functions remain constant 
within a given `context` also applies to location context. The function

> `counter.display()` is an example for this behavior. Below, read A will 

Reviewer 2: ```suggestion
[`counter.display`] is an example for this behavior. Below, read A will 
```

This makes it a link to the function's documentation page.

> `counter.display()` is an example for this behavior. Below, read A will 

Reviewer 2: ```suggestion
`counter.display()` is an example for this behavior. Below, Read A will 
```

Same below.
